### ISSUE TITLE
Fast-LoRA autograd, correctness half: hand-written backward for the LoRA projections, behind training.fast_lora

### ISSUE THREAD

**Issue by MakazhanAlpamys** [MakazhanAlpamys/Soup#792]
## Why

Every adapted projection in a Soup LoRA run is a peft `lora.Linear` (or `lora.bnb.Linear4bit` under `quantization: 4bit`), and its forward is generic autograd over separate ops. peft 0.20.0, `peft/tuners/lora/layer.py:1058`:

```

result = self.base_layer(x)

result = result + lora_B(lora_A(dropout(x))) * scaling

```

That is already the unmerged form `X W^T + s * (X A^T) B^T` — peft never materialises `W + s*BA`, so "stop merging the weight" is not the win here. What generic autograd cannot do is use structure *across* projections:

- q, k and v read the same `X`, yet run three rank-r GEMMs for `X A^T`, three for `dA`, and three for the A-side of `dX`.

- In a SwiGLU MLP, autograd keeps `silu(gate)` and the product `silu(gate) * up` alive for backward on top of `gate` and `up` themselves (the product because `down_proj`'s `lora_A` saves its input).

- The forward add, the `* scaling`, the 4-bit `result.clone()` (`peft/tuners/lora/bnb.py:533`) and the backward `dX` sums are each a separate elementwise kernel and allocation.

A `torch.autograd.Function` per block with a hand-written backward can concatenate the A matrices, accumulate `dX` into one buffer, and save only what backward needs. This is not new: Unsloth ships the design (`unsloth/kernels/fast_lora.py` — `LoRA_MLP`, `LoRA_QKV`, `LoRA_W`, Apache-2.0), and RunLoRA (arXiv:2312.03415, Cherniuk, Mikhalev, Oseledets, "Run LoRA Run: Faster and Lighter LoRA Implementations") enumerates forward/backward graphs for a LoRA layer, chooses between them per shape, and drops `XA` from the saved context. Liger Kernel, which Soup already wraps (`use_liger`), has no LoRA-aware op at time of writing, as far as we know: its `src/liger_kernel/ops/` at `main` (`95b01e9`, 2026-09-10) contains none, and `LigerSwiGLUMLP.forward` calls `self.down_proj(...)` around the PEFT-wrapped projections (`src/liger_kernel/transformers/swiglu.py:35`) without seeing the adapters.

This issue tracks the **correctness half only**: a kernel that computes the same math with the same trainable parameters, proven against the peft reference, plus a single-layer micro-benchmark. Whether it is faster end-to-end than anything else needs a larger Linux box and is separate work.

## What this is not

- **Not LoRA-FA (#725).** LoRA-FA is an optimizer that freezes A and trains B: a different recipe with a different result. This keeps A and B trainable, computes the same gradients, and should reproduce a plain LoRA loss curve. The two are orthogonal (a frozen-A variant would simply skip `dA`); composing them is out of scope.

- **Not new Triton kernels.** The first cut is plain PyTorch (`matmul` / `addmm` / elementwise). The maintainer's Windows box has no Triton installed at all.

## Scope

In:

- [ ] LoRA MLP path (gate/up/down around SwiGLU) — separate issue

- [ ] LoRA QKV path (q/k/v sharing one X) — separate issue

- [ ] LoRA single-projection path (o_proj, and the fallback for any other adapted Linear) — separate issue

- bf16 / fp16 / fp32 bases and NF4 bases; resident and layer-streamed models; with and without gradient checkpointing; `task: sft` first.

Out: DoRA, VeRA, 8-bit (`Linear8bitLt` has its own backward), mxfp4, `target_parameters` (raw 3-D tensors), MoE experts, multi-adapter mixed batches, MLX / unsloth backends, `torch.compile`, and any end-to-end tok/s or VRAM claim against Liger, Unsloth or HF.

## Design sketch

**Flag.** `training.fast_lora: bool = False`, with a SoupConfig cross-validator in the style of the existing ones (the dropout precedent is `LoraConfig._validate_target_parameter_compat`, `config/schema.py:287-291`; the footgun pattern is `schema.py:5146-5166`). Refuse, naming the reason: backend other than transformers; modality other than text; task other than sft (widen per trainer once tested); `lora.r == 0`; `use_dora` / `use_vera`; `target_parameters`; quantization other than `none` / `4bit`; and **`lora.dropout != 0`**. peft gives every projection its own dropout module (`layer.py:252-254`), so q/k/v and gate/up would see different `X_d` and the shared-X concatenation stops applying; Unsloth has the same condition (`unsloth/models/llama.py:3686`). Soup's default is 0.05 (`schema.py:72`), so the error must say that and name `lora.dropout: 0`.

**`use_liger`.** `utils/liger.py:75` calls `_apply_liger_kernel(model_type)` with defaults, and for Llama `swiglu=True` is the default, so the MLP class is already swapped. The MLP patch replaces the forward again on the instance. Refuse the pair or override and announce it — never silently.

**Hook points.** Resident: `apply_post_lora_patches` (`utils/peft_wiring.py:153`), which every wrapper calls right after `get_peft_model` (`trainer/sft.py:1530-1532`, `trainer/dpo.py:332-334`, ...) and after `prepare_model_for_kbit_training` (`sft.py:1394-1395`). Streamed: adapters attach at `utils/layer_stream_runtime.py:2326` and leave meta at `:2327`; the streaming setup builds its own `LoraConfig` (`trainer/stream_setup.py:814`) and does not call `apply_post_lora_patches` today, so the hook has to be added there after materialisation.

**Mechanism.** Instance-level `forward` replacement — what `install_dequant_forward` already does inside streamed layers (`layer_stream_runtime.py:228`) and what already works under `functional_call` (`:1040`). The patched forward reads the base weight at call time: `functional_call` substitutes streamed weights only for the duration of the call, so a reference captured at patch time is a meta placeholder. It delegates to the original peft forward whenever `disable_adapters`, `merged` or `adapter_names` is set (`peft/tuners/lora/layer.py:1035-1042`) — DPO/KTO's reference pass runs through `disable_adapters`, and a patch that ignores it makes the reference silently include the adapter. The patcher returns counts per path (as `install_dequant_forward` does); setup prints them and refuses if the flag is on and nothing was patched. On Llama, `target_modules: auto` resolves to peft's default `["q_proj", "v_proj"]` (`peft/utils/constants.py:83`), so on a default config only the QKV path, with k unadapted, can engage.

**The #331 rule.** Every tensor the backward reads goes through `ctx.save_for_backward`; nothing tensor-valued is stored as a ctx attribute. bitsandbytes stashes the packed weight as `ctx.tensors = (None, B)` (`bitsandbytes/autograd/_functions.py:358`, bnb 0.50.2) and dequantises it in backward (`:384`). Checkpointing cannot see that, and under layer streaming the reference aliases a pool slot that has been refilled: a bit-exact forward and wrong gradients on every layer but the last `stream_buffers` (#331, `layer_stream_runtime.py:161-184`). On ordinary QLoRA the same stash is harmless (`benchmarks/harness/issue331_qlora_scope.py`), which is exactly why a resident-only test would not catch a regression. For NF4, the packed weight *and* its quant-state tensors (absmax, nested absmax/offset — also pool buffers when streamed, `:124-131`) are saved and re-dequantised in backward with `dequantize_4bit`, mirroring the dtype handling of the #331 repair (`:203-223`). Soup checkpoints non-reentrantly on both paths (`utils/gradient_ckpt.py:89`, `layer_stream_runtime.py:1014-1016`), so saved tensors are dropped and recomputed. Backward is `@once_differentiable`, so double backward raises instead of returning garbage.

Code lives in `src/soup_cli/utils/fast_lora.py` with torch imported inside functions (`tests/test_cli_startup_is_light.py`).

## Acceptance criteria (shared by the three paths)

- [ ] `training.fast_lora` plus validators; one test per refusal asserting the reason keyword.

- [ ] `torch.autograd.gradcheck` in float64 on CPU for every Function, w.r.t. X and every A and B, base weight frozen. Runs in CI.

- [ ] Forward parity and backward parity against the unpatched peft modules, reported **separately**. fp32: `torch.testing.assert_close` at defaults. bf16/fp16: bit-exactness is not expected — concatenated A, `addmm` accumulation and scaling folded into the GEMM change the reduction order. Criterion: against a float64 reference, the fast path's max-abs error may not exceed the peft path's by more than 2x (proposed; the measured ratio is reported, not just pass/fail).

- [ ] NF4 parity, reference = unpatched peft over the same `Linear4bit`, never bf16.

- [ ] Gradients identical with and without `checkpoint(use_reentrant=False)` around the block.

- [ ] Streamed tiny model with more layers than `stream_buffers`, bf16 and NF4: LoRA gradients on **every** layer match streamed-without-fast_lora. Same path on both sides, because streamed NF4 vs resident NF4 is not bit-exact on sm_120 (#776).

- [ ] With `disable_adapters`, output equals the base-only output.

- [ ] Tiny model (SmolLM2-135M or a random tiny Llama), dropout 0, 50 steps: every step's loss equal to 3 decimals, fp32 on CPU in CI; bf16 on CUDA reported.

- [ ] Bytes saved for backward per block via `torch.autograd.graph.saved_tensors_hooks` (deduplicated by storage), fast vs plain, reported — no figure asserted in advance.

- [ ] `docs/peft-and-efficiency.md` states the correctness scope and that no speedup is claimed.

## Hardware needed

- CPU: gradcheck, fp32 parity, loss curve (all CI).

- Any CUDA GPU with bitsandbytes: bf16/fp16/NF4 parity, streaming, micro-benchmarks. GitHub runners have no GPU, so those tests skip in CI; post results with card, driver, SM clock and torch/bnb/peft versions. The maintainer's RTX 5070 Laptop 8 GB (Blackwell sm_120) can run all of it.

- A >=16 GB Linux box for any end-to-end comparison — not part of this.

## Honest limits

- The only performance number this produces is a single-layer micro-benchmark on a named card. Nothing here supports an end-to-end claim.

- Under gradient checkpointing — always on in layer streaming — a block's saved tensors exist only during its recompute, so a saving lowers a per-layer transient, not the stored total.

- Requiring dropout 0 changes the user's regularisation (not the kernel's math). The flag is useless on a default config until they opt into that.

- RunLoRA's point cuts both ways: the best graph depends on shape, and one fixed graph will not be best everywhere. A micro-benchmark showing no gain for a path is a valid result.

---

**Related, filed the same day:**

- #837 — Fast-LoRA autograd: fused LoRA MLP block (gate/up/down around SwiGLU) with a hand-written backward
- #838 — Fast-LoRA autograd: LoRA on q/k/v with one shared X (concatenated A, one dX accumulation)
- #839 — Fast-LoRA autograd: single-projection LoRA Function (o_proj, and the fallback for any other adapted Linear)
